Add quiz pagination support for Learning Mode.

dadish opened the pull request:
Fixes #4552
Fixes #4570

### Changes proposed in this Pull Request

* Introduces `sensei-lms/quiz-pagination` block that renders quiz pagination.

### Testing instructions

<!--
If the functionality of this PR is behind a feature flag, please include the
following testing step, using the correct name for the feature flag. (If you
aren't sure, just ignore this step)

* Enable the feature flag: `add_filter( 'sensei_feature_flag_{THE_FLAG_NAME}', '__return_true' );`
-->
* Enable the feature flag: `add_filter( 'sensei_feature_flag_course_theme', '__return_true' );`
* Enable the feature flag: `add_filter( 'sensei_feature_flag_quiz_pagination', '__return_true' );`
* Enable Course Theme / Learning Mode for a course.
* Create a lesson with a quiz.
* Set pagination for the quiz.
* Visit the lesson page and take a quiz.
* Confirm the pagination works as expected.
* Make sure to test with different numbers per page. (1 question per page, 3 question per page...)

<!--
Helpful tips for screenshots:
https://en.support.wordpress.com/make-a-screenshot/
-->
### Screenshot / Video


https://user-images.githubusercontent.com/2578542/148762086-06d45dc8-08a2-418e-869c-60bcc7c06f33.mp4

aaronfc commented on includes/blocks/course-theme/class-quiz-content.php:
```diff
@@ -64,6 +64,7 @@ private static function render_questions_loop() {
 			<?php
 		}
 
+		remove_action( 'sensei_single_quiz_questions_after', array( 'Sensei_Quiz', 'the_quiz_pagination' ), 9 );
```
Don't understand this. Is this something legacy that we have to stop doing? Why is it being removed "always" ?

dadish replied:
So the quiz pagination is automatically appended at the end of the post content. But in Learning Mode we need it in a separate place, in the footer tab (see below). The method `render_questions_loop` and the whole class `\Sensei\Blocks\Course_Theme\Quiz_Content` for that matter, is only rendered in the Learning Mode. Which means here we prevent the quiz pagination from being automatically appended into the post content and render it manually later where we need it to be.

I guess you're right. This line of code does not make sense without the context. I added a comment to clarify why it's there. cd4aebb2252c2b87bdc4deda9c695ef6c6cc58ed

aaronfc replied:
Thanks for the detailed explanation, @dadish 👍

aaronfc reviewed (commented):
I found something that might not be directly related to your change: the action buttons (next, previous and also complete) for the quiz do not work if you click outside the text.
![image](https://user-images.githubusercontent.com/799065/148777944-191f71cf-2f15-4c24-bab2-9b26bbefbe1b.png)

renatho commented on includes/blocks/course-theme/class-quiz-pagination.php:
```diff
@@ -0,0 +1,51 @@
+		Sensei_Quiz::the_quiz_pagination();
+		$pagination = ob_get_clean();
+		return ( "
+			<form method='POST' enctype='multipart/form-data'>
```
Since we're adding it to the `single.php` template, probably we need some check to not write this form on the lesson page, right?

dadish replied:
Not really. The quiz pagination works as a submit buttons inside a form. Each page button is a form value. If we remove form then quiz pagination will stop working. We ignore the form values if the quiz is already submitted. https://github.com/Automattic/sensei/blob/9ae94fa0fda3e6512b2fb09e58e800831fb6931f/includes/class-sensei-quiz.php#L461

renatho replied:
I mean that we're rendering the form on the lesson page, without having the quiz pagination.

<img width="1085" alt="Screen Shot 2022-01-11 at 18 36 47" src="https://user-images.githubusercontent.com/876340/149025467-15be9038-7264-403a-bcd8-50b127470569.png">

dadish replied:
Ohh, of course! 🤦 Fixed. 43de66e2c956a5ef56656f54c8921b1e47faae28

renatho reviewed (commented):
When accessing a lesson page, I see this in the logs:
> PHP Notice:  Trying to access array offset on value of type null in /Users/renatho/Workspace/automattic/Sensei/sensei/includes/class-sensei-quiz.php on line 1623

Also, if the pagination is set to "Single page", it's not loading anything. Is it something that is in work and will be part of the `Sensei_Quiz::the_quiz_pagination`, or should we handle this somehow?

dadish commented:
> I found something that might not be directly related to your change: the action buttons (next, previous and also complete) for the quiz do not work if you click outside the text.
Fixed. 1b2052058c513a10f2ad2a4bb13f415b39f7758f

renatho commented on themes/sensei-course-theme/single.php:
```diff
@@ -68,6 +68,7 @@
 
 		<!-- wp:sensei-lms/course-theme-notices /-->
 		<!-- wp:sensei-lms/course-content /-->
+		<!-- wp:sensei-lms/quiz-pagination /-->
```
Checking the designs, I think we don't want the actions here. Have you added it here for some reason?

dadish replied:
From the designs. The only one with quiz feedback is this one. I assume the action buttons were supposed to be at the end of the quiz. Which is where I put them. Could you point me to the correct place where I can see the correct designs, please.

<img width="391" alt="Screen Shot 2022-01-12 at 5 52 34 PM" src="https://user-images.githubusercontent.com/2578542/149153074-5233b402-9451-4670-883c-cf2537bb8975.png">

renatho replied:
Here you can check the designs: o2Xnqd2zgjaMaJFPmvNOx3-fi-20008411%3A49885

I forgot about the discussion on the slack channel that we want the pagination on the quiz results page too. But it's worth to check how it's there in the case of quiz without pagination. Maybe shouldn't we show the "Reset" link in the quiz results page? Or we could think in a better presentation for that.

<img width="976" alt="Screen Shot 2022-01-12 at 11 02 46" src="https://user-images.githubusercontent.com/876340/149154817-d1c98faf-83b6-4fdb-945b-ff68bdac4215.png">

dadish replied:
> I forgot about the discussion on the slack channel that we want the pagination on the quiz results page too. But it's worth to check how it's there in the case of quiz without pagination. Maybe shouldn't we show the "Reset" link in the quiz results page? Or we could think in a better presentation for that.

It looks a bit out of place when it's alone, I agree. Probably need a better copy, like "Reset Quiz" or "Restart Quiz". We already have a "Restart Quiz" at the top, so maybe we don't need it there. Do we also remove it when the quiz results are with pagination also? Or we show "Reset" link with pagination, but hide when there is no pagination?...

From the blocks/FSE perspective: This is now a "Quiz Actions" block. It's probably better if it does not have varying behavior depending on the context. It would be confusing for the editor/teacher to put this block at the end of quiz results and see it showing nothing at all on the frontend.

Either way, let me know how you want to proceed. I will make adjustments accordingly.

renatho replied:
Hum. I'd say that keeping it on the bottom of the results page, similar to the quiz page, would be better (also to keep consistency). I think it would seem less "lost" when it has only the "Reset" link. It would also mean adding the footer to this template too. But we don't want it on the lesson page. 🤔

Any idea on that?

renatho replied:
I think if you don't think of an easy solution for that, after checking the Miroslav issue, you can merge this PR and you can think about that in a new PR. =)

renatho reviewed (commented):
> @renatho
> 
> > When accessing a lesson page, I see this in the logs:
> > > PHP Notice:  Trying to access array offset on value of type null in /Users/renatho/Workspace/automattic/Sensei/sensei/includes/class-sensei-quiz.php on line 1623
> 
> I could not reproduce this one. Did I fix it somehow inadvertently?
> 
> > Also, if the pagination is set to "Single page", it's not loading anything. Is it something that is in work and will be part of the `Sensei_Quiz::the_quiz_pagination`, or should we handle this somehow?
> 
> Can't reproduce this either. 🤷

It seems both issues were solved with the last commits. I don't see the issues anymore. =)
One thing that I noticed is when it doesn't have pagination, the buttons are reversed.

Without pagination:
<img width="239" alt="Screen Shot 2022-01-12 at 10 25 42" src="https://user-images.githubusercontent.com/876340/149149194-53c724e8-9f5b-44e4-b644-f4c4901995f1.png">

With pagination:
<img width="226" alt="Screen Shot 2022-01-12 at 10 28 09" src="https://user-images.githubusercontent.com/876340/149149271-26a9339a-0bbf-481f-b211-ce55a62c7849.png">

Also, I wonder if the block name makes sense. If it includes the actions, probably it should be something more generic. Maybe "Quiz Actions" is more generic?

And I noticed that the save is not working yet. In the [issue](https://github.com/Automattic/sensei/issues/4552), you can find the possible reason for that:

> Also, notice that `sensei_single_quiz_content_inside_before` is part of the Single Quiz template in the normal mode, so `Sensei_Quiz::user_save_quiz_answers_listener` won't be called to save the quiz. We also need to fix it.

dadish commented:
> One thing that I noticed is when it doesn't have pagination, the buttons are reversed.

Fixed. 89191b75253ce500fe75ef3388c0368a826847b4

> Also, I wonder if the block name makes sense. If it includes the actions, probably it should be something more generic. Maybe "Quiz Actions" is more generic?

Updated. 43811bb2f18ed01767f37baccad9de5ddd9d9d5d


> And I noticed that the save is not working yet. In the [issue](https://github.com/Automattic/sensei/issues/4552), you can find the possible reason for that:

Fixed. 73f7c6ad1535234b027d4e39b32e56a15b030baf

dadish commented:
Btw @renatho, it seems like I also fixed #4570 when I addressed the "Save" button issue. I did a little refactoring to place the hooks in proper places and it looks a bit better now. LMKWYT 👍

aaronfc reviewed (approved):
Nothing else to add from my part.

renatho commented on assets/css/sensei-course-theme/quiz.scss:
```diff
@@ -61,13 +61,17 @@
-			flex: 1;
-
+		#sensei-quiz-actions {
+			flex-direction: row-reverse;
```
It sounds a little "hacky". I started a discussion (p1642194430020300-slack-C07418EJ0) to see if we can define a standard for that.

Also, I was checking that for mobile it's a little different from the design. Could you tweak it to be like the design in the different variations?

dadish replied:
> It sounds a little "hacky". I started a discussion (p1642194430020300-slack-C07418EJ0) to see if we can define a standard for that.

Got it. Will leave the current implementation as is until we confirm the correct behavior.
 
> Also, I was checking that for mobile it's a little different from the design. Could you tweak it to be like the design in the different variations?

Fixed. b3da657fe1fdd5a56142153608956fe4d97da326 But will probably need another update if the pagination rendering markup changes as per the above issue.

renatho reviewed (commented):
I'm also requesting a review for @m1r0 that worked on the pagination to see if he could make a quick check, or have any suggestions. =)

yscik commented on includes/blocks/course-theme/class-quiz-content.php:
```diff
@@ -25,32 +25,39 @@ class Quiz_Content {
-		\Sensei_Quiz::start_quiz_questions_loop();
+		ob_start();
+
+		do_action( 'sensei_single_quiz_content_inside_before', get_the_ID() );
```
That block does run most of the other hooks already, it's just a wrapper around the existing questions loop rendering via PHP templates. We do have [plans](https://github.com/Automattic/sensei/issues/4614) to move that to a more block-based approach, but I wonder if we still want to keep most of the quiz and question hooks, both for compatibility and extensibility. 
But the quiz save listener should definitely move out of any template/block filters, we can just add that to `init`/`template_redirect` like we did for lesson requests.

dadish replied:
I think it's best if we use the same [quiz template](https://github.com/automattic/sensei/blob/master/templates/single-quiz.php) that is used in regular mode to render the quiz in Learning Mode. So all the action hooks are the same in regular mode and in Learning Mode. This way we make sure the quiz behaves the same in regular mode and in learning mode. And we don't need to worry about two different cases when we extend quiz functionality with hooks.

The only difference would be is that in Learning Mode the hooks would swallow the buffer output to prevent attached actions from rendering custom markup during the hook execution. Because Learning Mode is all about Sensei's opinionated view on how the experience should look like and all the required markup is rendered by the Sensei Theme already.

yscik commented on includes/blocks/course-theme/class-quiz-actions.php:
```diff
@@ -0,0 +1,73 @@
+		}
+
+		return ( "
+			<form method='POST' enctype='multipart/form-data'>
+				{$actions}
+			</form>
```
Just having the `<form>` around the buttons means the submission won't include the actual answers, since those inputs are outside the form. It only works now because there is a form wrapper in the quiz template already. But then the browser will just ignore this form, since it's not valid to nest them.

I've been looking at replacing the form in the quiz template, and think we should just add a global form from filters if needed, instead of it being a dedicated wrapper block that can be removed to break thinks. 

Here is the code, it might be better suited for this PR and in the quiz-actions block:

```php
/**
	 * Add a form wrapper for quiz pages.
	 *
	 * @access private
	 */
	private function maybe_add_form_wrapper() {

		if ( ! has_block( 'sensei-lms/quiz-actions' ) ) {
			return;
		}

		add_action( 'wp_body_open', [ $this, 'start_form' ] );
		add_action( 'wp_footer', [ $this, 'end_form' ] );
	}

	/**
	 * Add a form open tag.
	 *
	 * @access private
	 */
	public function start_form() {
		echo '<form method="post" enctype="multipart/form-data" class="sensei-form">';
	}

	/**
	 * Add a form close tag.
	 *
	 * @access private
	 */
	public function end_form() {
		echo '</form>';
	}
```

dadish replied:
> Just having the `<form>` around the buttons means the submission won't include the actual answers, since those inputs are outside the form.

Nice catch! Don't know what I was thinking.

> instead of it being a dedicated wrapper block that can be removed to break thinks.

You mean user can potentially remove it in FSE mode?

> I've been looking at replacing the form in the quiz template, and think we should just add a global form from filters if needed,

Hmm. But this way it will wrap everything, including the admin bar. Not sure if that's safe.

dadish replied:
I just updated the logic for quiz actions. Now it is rendered inside a <form> if it is not already a part of a larger quiz form. 93a028e578b4baa26755249ace50890d162dfb9e

yscik replied:
Yeah if we rely on a 'Sensei Quiz Form' or similar block to be needed around the content and action blocks, it could become pretty unreliable, (Eg users removing it, or starting a template from scratch in the future).

I'll just handle this then in my PR for moving that top-level form out from the template.

dadish replied:
Sounds good.

yscik reviewed (dismissed):
Works well!

m1r0 commented:
I had a very strange bug when using the "Category questions" block. Here's how to reproduce it:

* Make 3 questions that are in the same questions category.
* Make a lesson with one "Category questions" block.
* Set the number of questions for that block to 2.
* Enable the quiz "random questions order" setting.
* Enable the pagination and set it to 1 question per page.

With these settings, we expect 2 random questions pulled from the selected category (1 per page, a total of 2 pages).
The issue is that after the first page is submitted, **_sometimes_** another question is pulled from that category, and the questions become 3. I couldn't reproduce this outside of this branch.

https://user-images.githubusercontent.com/1612178/149994228-b619f70f-5c6c-417a-8b3e-cd3b57523fae.mp4

m1r0 commented on assets/css/frontend.scss:
```diff
@@ -455,6 +455,10 @@ a.sensei-certificate-link {
 		margin: 0;
+		padding: 0;
+		button.button {
+			padding: 8px 11px;
```
I think this will also change the button padding outside of the course theme. Is this intentional?

dadish replied:
Nice catch! I put it away into course-theme related CSS. 9d5360a3562dacdb5f96b5e81d007b032bd889ce

dadish commented:
> The issue is that after the first page is submitted, sometimes another question is pulled from that category, and the questions become 3. I couldn't reproduce it on master.

Thank you for catching that @m1r0! Fixed here. 02aca5e097e03b68db4effbb2a595f54bf7a5ada

m1r0 commented:
A bit late to the party, but all works great!

dadish commented:
> A bit late to the party, but all works great!

Thanks a lot for taking another look @m1r0! Appreciate!
